sql: distsql plan node

RaduBerinde opened the pull request:
Creating a plan node that can interface between the existing plan tree and the
new distSQL flow.

Adding code to convert a scanNode to a local flow with a single tableReader. For
now this is experimental code, controlled by a global constant.

I temporarily flipped the switch and ran some preliminary benchmarks. The
results aren't bad for large scans but we get disconcerting regressions for the
small cases. I spent about a day fiddling with Scan1; profiling traces show the
channel operation taking a lot of time (and yet if I remove the fetch, this
delay goes away; and if I remove the channel altogether and just wait for the
waitgroup, the delay is still there). My conclusions so far:
- the KV fetch, which is the exact same operation the scanNode does is taking
  40-60us more just because we are doing it in a different goroutine
- the first channel operation (even if we don't send the row and just do the
  close) sometimes takes tens of microseconds during which we are spinning on
  some internal go futex

I tried to make a simple test that simulates the same goroutines and channels
but could not reproduce. It feels like some kind of scheduling issue or
interaction with something specific to the activity generated by the KV fetch
operation.

```
Scan1_Cockroach-16                    195µs ± 4%   369µs ± 3%   +89.57%  (p=0.000 n=20+17)
Scan10_Cockroach-16                   226µs ± 4%   412µs ± 4%   +82.10%  (p=0.000 n=20+19)
Scan100_Cockroach-16                  480µs ± 4%   736µs ± 4%   +53.30%  (p=0.000 n=20+20)
Scan1000_Cockroach-16                2.73ms ± 3%  3.17ms ± 4%   +16.29%  (p=0.000 n=19+18)
Scan10000_Cockroach-16               24.8ms ± 7%  27.4ms ± 4%   +10.46%  (p=0.000 n=20+17)
Scan1000Limit1_Cockroach-16           202µs ± 4%   385µs ± 5%   +90.32%  (p=0.000 n=20+19)
Scan1000Limit10_Cockroach-16          234µs ± 4%   426µs ± 5%   +82.32%  (p=0.000 n=20+20)
Scan1000Limit100_Cockroach-16         485µs ± 5%   735µs ± 4%   +51.62%  (p=0.000 n=19+19)
Scan10000FilterLimit1_Cockroach-16    394µs ± 4%   959µs ± 4%  +143.38%  (p=0.000 n=18+20)
Scan10000FilterLimit10_Cockroach-16   443µs ± 4%  1012µs ± 7%  +128.55%  (p=0.000 n=19+20)
Scan10000FilterLimit50_Cockroach-16  2.59ms ± 7%  3.05ms ±12%   +17.38%  (p=0.000 n=18+20)
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7160)

<!-- Reviewable:end -->

andreimatei commented:
If we insist on it, we can eliminate the goroutine for very simple point lookup queries, right? We can have the consumer (`DistSqlNode`) call `Run` on the `tableReader`, replace the `RowChannel` with an implementation of a `RowReceiver` that takes a continuation to run when it's closed (which would be most of `DistSqlNode.Next()`) .

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> #### sql: distsql plan node
> 
> Creating a plan node that can interface between the existing plan tree and the
> 
> new distSQL flow.
> 
> 
> 
> Adding code to convert a scanNode to a local flow with a single tableReader. For
> 
> now this is experimental code, controlled by a global constant.
> 
> 
> 
> I temporarily flipped the switch and ran some preliminary benchmarks. The
> 
> results aren't bad for large scans but we get disconcerting regressions for the
> 
> small cases. I spent about a day fiddling with Scan1; profiling traces show the
> 
> channel operation taking a lot of time (and yet if I remove the fetch, this
> 
> delay goes away; and if I remove the channel altogether and just wait for the
> 
> waitgroup, the delay is still there). My conclusions so far:
> 
>  - the KV fetch, which is the exact same operation the scanNode does is taking
> 
>    40-60us more just because we are doing it in a different goroutine
> 
>  - the first channel operation (even if we don't send the row and just do the
> 
>    close) sometimes takes tens of microseconds during which we are spinning on
> 
>    some internal go futex
> 
> 
> 
> I tried to make a simple test that simulates the same goroutines and channels
> 
> but could not reproduce. It feels like some kind of scheduling issue or
> 
> interaction with something specific to the activity generated by the KV fetch
> 
> operation.
> 
> 
> 
> ```
> 
> Scan1_Cockroach-16                    195µs ± 4%   369µs ± 3%   +89.57%  (p=0.000 n=20+17)
> 
> Scan10_Cockroach-16                   226µs ± 4%   412µs ± 4%   +82.10%  (p=0.000 n=20+19)
> 
> Scan100_Cockroach-16                  480µs ± 4%   736µs ± 4%   +53.30%  (p=0.000 n=20+20)
> 
> Scan1000_Cockroach-16                2.73ms ± 3%  3.17ms ± 4%   +16.29%  (p=0.000 n=19+18)
> 
> Scan10000_Cockroach-16               24.8ms ± 7%  27.4ms ± 4%   +10.46%  (p=0.000 n=20+17)
> 
> Scan1000Limit1_Cockroach-16           202µs ± 4%   385µs ± 5%   +90.32%  (p=0.000 n=20+19)
> 
> Scan1000Limit10_Cockroach-16          234µs ± 4%   426µs ± 5%   +82.32%  (p=0.000 n=20+20)
> 
> Scan1000Limit100_Cockroach-16         485µs ± 5%   735µs ± 4%   +51.62%  (p=0.000 n=19+19)
> 
> Scan10000FilterLimit1_Cockroach-16    394µs ± 4%   959µs ± 4%  +143.38%  (p=0.000 n=18+20)
> 
> Scan10000FilterLimit10_Cockroach-16   443µs ± 4%  1012µs ± 7%  +128.55%  (p=0.000 n=19+20)
> 
> Scan10000FilterLimit50_Cockroach-16  2.59ms ± 7%  3.05ms ±12%   +17.38%  (p=0.000 n=18+20)
> 
> ````
> 
> 
> 
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/dist_sql_node.go, line 156 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7160#-KKAFysxZfW6Cij_H4BK:-KKAFysxZfW6Cij_H4BL:649660670) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b08cdcbc82691dfbcb41b03746853a53e4ef763/sql/dist_sql_node.go#L156)):*
> ```Go
> 		n.resultColumns = make([]ResultColumn, len(tmp))
> 		for i, orig := range tmp {
> 			n.resultColumns[i].Name = fmt.Sprintf("$%d", i)
> ```

lolol I don't know what this does. Can you expand on the comment above?

---

*[sql/executor.go, line 945 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7160#-KKADHzPGLVhUtaMJFn-:-KKADHzPGLVhUtaMJFn0:-889325761) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b08cdcbc82691dfbcb41b03746853a53e4ef763/sql/executor.go#L945)):*
> ```Go
> 	}
> 
> 	// TODO(radu): experimental code for testing distSQL flows.
> ```

nit: move comment to const definition

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7160#-:-KKAF4wzT4kY36uCBoaJ:-323436241)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Would we buffer the results in memory somewhere and then return them one by one in each `Next` call? If yes, that would be pretty easy; and we could use the same model even if we are using an index - we would first run `tableReader`, get the results, and then run `joinReader`.

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> If we insist on it, we can eliminate the goroutine for very simple point lookup queries, right? We can have the consumer (`DistSqlNode`) call `Run` on the `tableReader`, replace the `RowChannel` with an implementation of a `RowReceiver` that takes a continuation to run when it's closed (which would be most of `DistSqlNode.Next()`) .
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/dist_sql_node.go, line 156 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7160#-KKAFysxZfW6Cij_H4BK:-KKALi2y_hRonh5qhkfY:1930683831) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b08cdcbc82691dfbcb41b03746853a53e4ef763/sql/dist_sql_node.go#L156)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> lolol I don't know what this does. Can you expand on the comment above?
</details>
Sure, will do.

The distsql layer uses the placeholder syntax `$x` to refer to columns (see https://github.com/cockroachdb/cockroach/blob/master/sql/distsql/expr.go#L50). We are making it print like that by temporarily renaming the columns to (literally) "$0", "$1", ..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7160#-:-KKAKb3nDa29C6uyey3b:-1058657809)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
that's what I was thinking. Or even just doing this when there's a single lookup by primary key.

This was also supposed to be a LGTM

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Would we buffer the results in memory somewhere and then return them one by one in each `Next` call? If yes, that would be pretty easy; and we could use the same model even if we are using an index - we would first run `tableReader`, get the results, and then run `joinReader`.
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7160#-:-KKAMU8ybMDcFssIpyGy:1163687617)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7160)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: distsql plan node

Creating a plan node that can interface between the existing plan tree and the
new distSQL flow.

Adding code to convert a scanNode to a local flow with a single tableReader. For
now this is experimental code, controlled by a global constant.

I temporarily flipped the switch and ran some preliminary benchmarks. The
results aren't bad for large scans but we get disconcerting regressions for the
small cases. I spent about a day fiddling with Scan1; profiling traces show the
channel operation taking a lot of time (and yet if I remove the fetch, this
delay goes away; and if I remove the channel altogether and just wait for the
waitgroup, the delay is still there). My conclusions so far:
 - the KV fetch, which is the exact same operation the scanNode does is taking
   40-60us more just because we are doing it in a different goroutine
 - the first channel operation (even if we don't send the row and just do the
   close) sometimes takes tens of microseconds during which we are spinning on
   some internal go futex

I tried to make a simple test that simulates the same goroutines and channels
but could not reproduce. It feels like some kind of scheduling issue or
interaction with something specific to the activity generated by the KV fetch
operation.

```
Scan1_Cockroach-16                    195µs ± 4%   369µs ± 3%   +89.57%  (p=0.000 n=20+17)
Scan10_Cockroach-16                   226µs ± 4%   412µs ± 4%   +82.10%  (p=0.000 n=20+19)
Scan100_Cockroach-16                  480µs ± 4%   736µs ± 4%   +53.30%  (p=0.000 n=20+20)
Scan1000_Cockroach-16                2.73ms ± 3%  3.17ms ± 4%   +16.29%  (p=0.000 n=19+18)
Scan10000_Cockroach-16               24.8ms ± 7%  27.4ms ± 4%   +10.46%  (p=0.000 n=20+17)
Scan1000Limit1_Cockroach-16           202µs ± 4%   385µs ± 5%   +90.32%  (p=0.000 n=20+19)
Scan1000Limit10_Cockroach-16          234µs ± 4%   426µs ± 5%   +82.32%  (p=0.000 n=20+20)
Scan1000Limit100_Cockroach-16         485µs ± 5%   735µs ± 4%   +51.62%  (p=0.000 n=19+19)
Scan10000FilterLimit1_Cockroach-16    394µs ± 4%   959µs ± 4%  +143.38%  (p=0.000 n=18+20)
Scan10000FilterLimit10_Cockroach-16   443µs ± 4%  1012µs ± 7%  +128.55%  (p=0.000 n=19+20)
Scan10000FilterLimit50_Cockroach-16  2.59ms ± 7%  3.05ms ±12%   +17.38%  (p=0.000 n=18+20)

````

RaduBerinde committed:
distsql: export StreamMsg members
